eth/downloader/queue.go

```diff
@@ -86,81 +131,141 @@ func (q *queue) Reset() {
-// pending and already downloaded.
-func (q *queue) Size() (int, int) {
+// InFlightBlocks retrieves whether there are block fetch requests currently in
+// flight.
+func (q *queue) InFlightBlocks() bool {
 	q.lock.RLock()
 	defer q.lock.RUnlock()
 
-	return len(q.hashPool) + len(q.headerPool), len(q.blockPool)
+	return len(q.blockPendPool) > 0
 }
 
-// Pending retrieves the number of blocks pending for retrieval.
-func (q *queue) Pending() int {
+// InFlightReceipts retrieves whether there are receipt fetch requests currently
+// in flight.
+func (q *queue) InFlightReceipts() bool {
 	q.lock.RLock()
 	defer q.lock.RUnlock()
 
-	return q.hashQueue.Size() + q.headerQueue.Size()
+	return len(q.receiptPendPool) > 0
 }
 
-// InFlight retrieves the number of fetch requests currently in flight.
-func (q *queue) InFlight() int {
+// InFlightNodeData retrieves whether there are node data entry fetch requests
+// currently in flight.
+func (q *queue) InFlightNodeData() bool {
 	q.lock.RLock()
 	defer q.lock.RUnlock()
 
-	return len(q.pendPool)
+	return len(q.statePendPool)+int(atomic.LoadInt32(&q.stateProcessors)) > 0
 }
 
-// Throttle checks if the download should be throttled (active block fetches
-// exceed block cache).
-func (q *queue) Throttle() bool {
+// Idle returns if the queue is fully idle or has some data still inside. This
+// method is used by the tester to detect termination events.
+func (q *queue) Idle() bool {
 	q.lock.RLock()
 	defer q.lock.RUnlock()
 
-	// Calculate the currently in-flight block requests
+	queued := q.hashQueue.Size() + q.blockTaskQueue.Size() + q.receiptTaskQueue.Size() + q.stateTaskQueue.Size()
+	pending := len(q.blockPendPool) + len(q.receiptPendPool) + len(q.statePendPool)
+	cached := len(q.blockDonePool) + len(q.receiptDonePool)
+
+	q.stateSchedLock.RLock()
+	if q.stateScheduler != nil {
+		queued += q.stateScheduler.Pending()
+	}
+	q.stateSchedLock.RUnlock()
+
+	return (queued + pending + cached) == 0
+}
+
+// ThrottleBlocks checks if the download should be throttled (active block (body)
+// fetches exceed block cache).
+func (q *queue) ThrottleBlocks() bool {
+	q.lock.RLock()
+	defer q.lock.RUnlock()
+
+	// Calculate the currently in-flight block (body) requests
 	pending := 0
-	for _, request := range q.pendPool {
+	for _, request := range q.blockPendPool {
 		pending += len(request.Hashes) + len(request.Headers)
 	}
-	// Throttle if more blocks are in-flight than free space in the cache
-	return pending >= len(q.blockCache)-len(q.blockPool)
+	// Throttle if more blocks (bodies) are in-flight than free space in the cache
+	return pending >= len(q.resultCache)-len(q.blockDonePool)
 }
 
-// Has checks if a hash is within the download queue or not.
-func (q *queue) Has(hash common.Hash) bool {
+// ThrottleReceipts checks if the download should be throttled (active receipt
+// fetches exceed block cache).
+func (q *queue) ThrottleReceipts() bool {
```

Comment:
It's name make it sound like it's doing something; perhaps call it `NeedToThrottleReceipts` or some such.